java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java

```diff
@@ -46,47 +49,69 @@
 @ActivatedByDefault
 @SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.UNDERSTANDABILITY)
 @SqaleConstantRemediation("5min")
-public class UnusedPrivateMethodCheck extends BytecodeVisitor {
+public class UnusedPrivateMethodCheck extends SubscriptionBaseVisitor {
 
-  private AsmClass asmClass;
+  private static final MethodInvocationMatcherCollection SERIALIZABLE_METHODS = MethodInvocationMatcherCollection.create(
+    MethodMatcher.create().name("writeObject").addParameter("java.io.ObjectOutputStream"),
+    MethodMatcher.create().name("readObject").addParameter("java.io.ObjectInputStream"),
+    MethodMatcher.create().name("writeReplace"),
+    MethodMatcher.create().name("readResolve"),
+    MethodMatcher.create().name("readObjectNoData")
+    );
 
   @Override
-  public void visitClass(AsmClass asmClass) {
-    this.asmClass = asmClass;
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.COMPILATION_UNIT);
   }
 
   @Override
-  public void visitMethod(AsmMethod asmMethod) {
-    if (isPrivateUnused(asmMethod) && !isExcludedFromCheck(asmMethod)) {
-      String messageStr = "Private method '" + asmMethod.getName() + "' is never used.";
-      if ("<init>".equals(asmMethod.getName())) {
-        messageStr = "Private constructor '" + asmClass.getDisplayName() + "(";
-        List<String> params = Lists.newArrayList();
-        for (Parameter param : MethodSignatureScanner.scan(asmMethod.getGenericKey()).getArgumentTypes()) {
-          String paramName = param.getClassName();
-          if (StringUtils.isEmpty(paramName)) {
-            paramName = MethodSignatureScanner.getReadableType(param.getJvmJavaType());
-          }
-          params.add(paramName + (param.isArray() ? "[]" : ""));
-        }
-        messageStr += Joiner.on(",").join(params) + ")' is never used.";
+  public void visitNode(Tree tree) {
+    for (Tree memberTree : ((CompilationUnitTree) tree).types()) {
+      MethodInvocationVisitor methodInvocationVisitor = new MethodInvocationVisitor();
+      memberTree.accept(methodInvocationVisitor);
+      memberTree.accept(new MethodVisitor(methodInvocationVisitor.calledPrivateMethods));
+    }
+  }
+
+  private static class MethodInvocationVisitor extends BaseTreeVisitor {
+    final Set<Symbol> calledPrivateMethods = new HashSet<>();
+
+    @Override
+    public void visitNewClass(NewClassTree tree) {
+      Symbol constructorSymbol = ((NewClassTreeImpl) tree).getConstructorIdentifier().symbol();
```

Comment:
@benzonico your opinion on this? I am tempted to say we should expose the symbol of the constructor through the interface. NewClassTree.constructor()?